src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/DefaultGeneratedDocumentPublisher.cs

```diff
@@ -123,6 +124,16 @@ public override void PublishCSharp(ProjectKey projectKey, string filePath, Sourc
             HostDocumentVersion = hostDocumentVersion,
         };
 
+        // HACK: We know about a document being in multiple projects, but despite having ProjectKeyId in the request, currently the other end
+        // of this LSP message only knows about a single document file path. To prevent confusing them, we just send an update for the first project
+        // in the list.
+        if (_projectSnapshotManager is not null &&
+            _projectSnapshotManager.GetLoadedProject(projectKey) is { } projectSnapshot &&
+            _projectSnapshotManager.GetAllProjectKeys(projectSnapshot.FilePath).First() != projectKey)
```

Comment:
Consider capturing `_projectSnapshotManager` in a local to defend against future changes that might change it (since it's nullable).

```suggestion
        if (_projectSnapshotManager is { } projectSnapshotManager
            projectSnapshotManager.GetLoadedProject(projectKey) is { } projectSnapshot &&
            projectSnapshotManager.GetAllProjectKeys(projectSnapshot.FilePath).First() != projectKey)
```

Reply:
The null check is only there for tests, but I like that suggestion anyway, just in case. In truth I'm beggining to really despise the way the project snapshot manager works, that it can't be injected via traditional DI mechanisms.